DownFall/TargetingViewModel.swift

```diff
@@ -0,0 +1,190 @@
+                InputQueue.append(Input(InputType.itemUseCanceled))
+            } else {
+                if oldValue == nil {
+                    InputQueue.append(Input(InputType.itemUseSelected(ability!)))
+                }
+                autoTarget()
+            }
+            updateCallback?()
+        }
+    }
+    
+    var numberOfTargets: Int {
+        return ability?.targets ?? 0
+    }
+    
+    var typesOfTargets: [TileType] {
+        return ability?.targetTypes ?? []
+    }
+    
+    var toastMessage: String {
+        if ability == nil {
+            return ""
+        }
+        let baseString = "Choose "
+        let types = Set<String>(self.typesOfTargets.map {
+            return $0.humanReadable
+        }).joined(separator: ",")
+        return baseString + "\(self.numberOfTargets) " + types
+    }
+    
+    var currentTargets: [Target] = [] {
+        didSet {
+            self.updateCallback?()
+            InputQueue.append(Input(InputType.itemCanBeUsed(legallyTargeted)))
+        }
+    }
+    
+    
+    var legallyTargeted: Bool {
+        return self.currentTargets.allSatisfy({
+            return $0.isLegal
+        }) && self.currentTargets.count == self.numberOfTargets
+    }
+    
+    
+    private var tiles: [[Tile]]? = nil {
+        didSet {
+            autoTarget()
+        }
+    }
+    
+    init() {
+        Dispatch.shared.register { [weak self] (input) in
+            switch input.type {
+            case .itemUseSelected(let ability):
+                self?.ability = ability
+                self?.updateCallback?()
+            case .transformation(let trans):
+                if let inputType = trans.inputType,
+                    case InputType.itemUseSelected(_) = inputType,
+                    let endTiles = trans.endTiles
+                {
+                    self?.tiles = endTiles
+                    self?.updateCallback?()
+                }
+                
+            case .itemUseCanceled:
+                ()
+            case .itemUsed:
+                self?.ability = nil
+            default:
+                ()
+            }
+        }
+    }
+    
+    private func isTargetLegal(_ coord: TileCoord) -> Bool {
+        guard let tiles = tiles else { return false }
+        //need to access what the actualy tile is
+        if typesOfTargets.contains(tiles[coord].type) {
+            return true
+        }
+        return false
+    }
+    
+    /**
+    Toggles targeted-ness of a tile.  If a tile is not targeted it becomes targeted.  The opposite is true.
+     
+     - Note: If the max number of targets for an ability is reached and a unique coord is passed in. The most recently placed target is un-targeted. And the passed in coord becomes targeted.
+     - Parameters coord: The coord that is being targetted
+     - Returns: Nothing
+     */
+    func didTarget(_ coord: TileCoord) {
+        guard numberOfTargets > 0 else { fatalError("Something funcky is happening") }
```

Comment:
this could just mean that we dont have an ability set- which also means we cant be targeting.  I think it makes sense to crash here so we can find bugs in our code.  Maybe just check for a nil ability instead tho